catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,370 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.content.monitor;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.File;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+import javax.annotation.Nullable;
+import org.apache.commons.io.FileUtils;
+import org.codice.ddf.catalog.content.monitor.synchronizations.CompletionSynchronization;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Based on {@link org.apache.commons.io.monitor.FileAlterationObserver}, except modified to only
+ * update the observer state on successful async request
+ *
+ * <p>This implementation only works with one AsyncFileAlterationListener.
+ *
+ * <p>Every time the AsyncFileAlterationObserver is polled by calling {@code checkAndNotify()}, the
+ * observer will check all the {@link File}'s and {@link AsyncFileEntry}'s that are under the
+ * directory being monitored, and call the {@link AsyncFileAlterationListener}'s corresponding
+ * methods
+ *
+ * @see AsyncFileAlterationListener
+ */
+public class AsyncFileAlterationObserver implements Serializable {
+
+  @Nullable private final AsyncFileEntry rootFile; //  In case GSON returns it.
+  @Nullable private transient AsyncFileAlterationListener listener = null;
+  private final transient Set<AsyncFileEntry> processing = new ConcurrentSkipListSet<>();
+  private static final Logger LOGGER = LoggerFactory.getLogger(AsyncFileAlterationObserver.class);
+  private final transient Object listenerLock = new Object();;
+
+  public AsyncFileAlterationObserver(File fileToObserve) {
+    if (fileToObserve == null) {
+      throw new IllegalArgumentException();
+    }
+    rootFile = new AsyncFileEntry(fileToObserve);
+  }
+
+  //  For GSON serialization;
+  private AsyncFileAlterationObserver() {
+    rootFile = null;
+  }
+
+  public void onLoad() {
```

Comment:
📘 Needed due to gson serialization. File's need to know their parent but gson will make that into an infinite loop, so those are nulled out before writing and have to be reinitialized when loaded. 

Probably should put a javadoc here.